File: interpreter-for-cpp/src/Runtime/Sentence/SentenceExpressionInside.cpp
#include "SentenceExpressionInside.h"
#include "../Value/ValueTool.h"
#include "SentenceExpressionVariable.h"

using namespace peak::interpreter;

SentenceExpressionInside::SentenceExpressionInside(std::shared_ptr<SentenceExpression> header, std::vector<std::shared_ptr<SentenceExpression>> insides)
	: _header(header), _insides(insides) {
}
ExecuteResult SentenceExpressionInside::Execute(std::shared_ptr<Space> space) {
	if (!Sentence::IsSuccess(_header->Execute(space))) {
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Inside, "The header expression execute failed!");
		return ExecuteResult::Failed;
	}
	auto headerValue = _header->GetValue();
	if (!ValueTool::IsObject(headerValue)) {
		ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Inside, "The header expression isn't a object!");
		return ExecuteResult::Failed;
	}
	auto tempValue = headerValue;
	for (auto expression : _insides) {
		if (!ValueTool::IsObject(tempValue)) {
			ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Inside, "The expression isn't a object!");
			return ExecuteResult::Failed;
		}
		auto objSpace = std::static_pointer_cast<ValueObject>(tempValue)->GetSpace();
		auto expType = expression->GetExpressionType();
		auto executeRet = ExecuteResult::Failed;
		if (expType == ExpressionType::Function) {
			executeRet = std::static_pointer_cast<SentenceExpressionFunctionCall>(expression)->ExecuteFromInside(objSpace, space);
		} else {
			executeRet = expression->Execute(objSpace);
		}
		if (!Sentence::IsSuccess(executeRet)) {
			ErrorLogger::LogRuntimeError(ErrorRuntimeCode::Inside, "The inside expression execute failed!");
			return ExecuteResult::Failed;
		}
		tempValue = expression->GetValue();
	}
	SetValue(tempValue);
	return ExecuteResult::Successed;
}

File: src/Runtime/Sentence/SentenceExpressionInside.cs
/*
	PeakScript: https://github.com/Jenocn/PeakScript
	By Jenocn: https://jenocn.github.io
*/

using System.Collections.Generic;

namespace peak.interpreter {

	public class SentenceExpressionInside : SentenceExpression {
		private SentenceExpression _header = null;
		private List<SentenceExpression> _insides = null;
		public SentenceExpressionInside(SentenceExpression header, List<SentenceExpression> insides) {
			_header = header;
			_insides = insides;
		}
		public override ExpressionType GetExpressionType() { return ExpressionType.Inside; }
		public override ExecuteResult Execute(Space space) {
			if (!Sentence.IsSuccess(_header.Execute(space))) {
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Inside, "The header expression execute failed!");
				return ExecuteResult.Failed;
			}
			var headerValue = _header.value;
			if (!ValueTool.IsObject(headerValue)) {
				ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Inside, "The header expression isn't a object!");
				return ExecuteResult.Failed;
			}
			var tempValue = headerValue;
			foreach (var expression in _insides) {
				if (!ValueTool.IsObject(tempValue)) {
					ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Inside, "The expression isn't a object!");
					return ExecuteResult.Failed;
				}
				var objSpace = (tempValue as ValueObject).space;
				var expType = expression.GetExpressionType();
				var executeRet = ExecuteResult.Failed;
				if (expType == ExpressionType.Function) {
					executeRet = (expression as SentenceExpressionFunctionCall).ExecuteFromInside(objSpace, space);
				} else {
					executeRet = expression.Execute(objSpace);
				}
				if (!Sentence.IsSuccess(executeRet)) {
					ErrorLogger.LogRuntimeError(ErrorRuntimeCode.Inside, "The inside expression execute failed!");
					return ExecuteResult.Failed;
				}
				tempValue = expression.value;
			}
			SetValue(tempValue);
			return ExecuteResult.Successed;
		}

	}

} // namespace peak.interpreter